src/fdc3/dotnet/DesktopAgent/src/MorganStanley.ComposeUI.DesktopAgent/Infrastructure/Internal/UserChannelSetReader.cs

```diff
@@ -0,0 +1,105 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * See the NOTICE file distributed with this work for additional information
+ * regarding copyright ownership. Unless required by applicable law or agreed
+ * to in writing, software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+using MorganStanley.ComposeUI.Fdc3.DesktopAgent.Exceptions;
+using System.Reflection;
+using System.Text.Json;
+using MorganStanley.ComposeUI.Fdc3.DesktopAgent.Protocol;
+using System.IO.Abstractions;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.Extensions.Options;
+using MorganStanley.ComposeUI.Fdc3.DesktopAgent.DependencyInjection;
+using MorganStanley.ComposeUI.Fdc3.DesktopAgent.Converters;
+using System.Text.Json.Serialization;
+using System.Collections.Immutable;
+
+namespace MorganStanley.ComposeUI.Fdc3.DesktopAgent.Infrastructure.Internal;
+
+internal class UserChannelSetReader : IUserChannelSetReader, IDisposable
+{
+    private readonly Fdc3DesktopAgentOptions _options;
+    private readonly IFileSystem _fileSystem;
+    private readonly HttpClient _httpClient;
+    private readonly ILogger<UserChannelSetReader> _logger;
+    private IReadOnlyDictionary<string, ChannelItem>? _userChannelSet;
+    private readonly JsonSerializerOptions _jsonSerializerOptions = new(JsonSerializerDefaults.Web)
+    {
+        Converters = { new DisplayMetadataJsonConverter(), new JsonStringEnumConverter() }
+    };
+
+    public UserChannelSetReader(
+        IOptions<Fdc3DesktopAgentOptions> options,
+        IFileSystem? fileSystem = null, 
+        ILogger<UserChannelSetReader>? logger = null)
+    {
+        _options = options.Value;
+        _fileSystem = fileSystem ?? new FileSystem();
+        _httpClient = new HttpClient();
+        _logger = logger ?? NullLogger<UserChannelSetReader>.Instance;
+    }
+
+    public void Dispose()
+    {
+        _httpClient.Dispose();
+    }
+
+    public async Task<IReadOnlyDictionary<string, ChannelItem>> GetUserChannelSet(CancellationToken cancellationToken = default)
+    {
+        if (_userChannelSet != null)
+        {
+            return _userChannelSet;
+        }
+
+        var uri = _options.UserChannelConfigFile;
+
+        if (uri == null && _options.UserChannelConfig == null)
+        {
+            var assembly = Assembly.GetExecutingAssembly();
+            using var stream = assembly.GetManifestResourceStream(ResourceNames.DefaultUserChannelSet);
+            if (stream != null)
+            {
+                using var streamReader = new StreamReader(stream);
+                var result = streamReader.ReadToEnd();
+                var userChannels = JsonSerializer.Deserialize<ChannelItem[]>(result, _jsonSerializerOptions);
```

Comment:
`JsonSerializer.Deserialize()` has an overload which accepts a stream - you don't need to allocate (and then throw away) a `string` here.